drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,1901 @@
+	}
+
+	/* Reset USB device address to 0 */
+	m_numaker_usbd_reset_addr(dev);
+}
+
+/* Interrupt bottom half processing for Setup/EP data transaction */
+static void numaker_usbd_ep_bh(struct numaker_usbd_ep *ep_cur,
+			       enum usb_dc_ep_cb_status_code status_code)
+{
+	struct usb_dc_numaker_device *dev = ep_cur->dev;
+	struct usb_dc_numaker_data *data = dev->data;
+	struct numaker_usbd_ep_mgmt *ep_mgmt = &data->ep_mgmt;
+
+	if (status_code == USB_DC_EP_SETUP) {
+		/* Zephyr USB device stack passes Setup packet via CTRL OUT EP. */
+		__ASSERT_NO_MSG(ep_cur->ep_addr == USB_EP_GET_ADDR(0, USB_EP_DIR_OUT));
+
+		if (numaker_usbd_ep_fifo_used(ep_cur)) {
+			LOG_WRN("New Setup will override previous Control OUT data");
+		}
+
+		/* We should have reserved 1st/2nd EP contexts for CTRL OUT/IN */
+		__ASSERT_NO_MSG(ep_cur->ep_addr == USB_EP_GET_ADDR(0, USB_EP_DIR_OUT));
+		__ASSERT_NO_MSG((ep_cur + 1)->ep_addr == USB_EP_GET_ADDR(0, USB_EP_DIR_IN));
+
+		/* Reset CTRL OUT/IN FIFO due to new Setup packet */
+		numaker_usbd_ep_fifo_reset(ep_cur);
+		numaker_usbd_ep_fifo_reset(ep_cur + 1);
+
+		/* Relinquish CTRL OUT/IN DMA buffer ownership on behalf of H/W */
+		numaker_usbd_ep_abort(ep_cur);
+		numaker_usbd_ep_abort(ep_cur + 1);
+
+		/* Mark new Setup packet for read */
+		m_numaker_usbd_setup_fifo_copy_to_user(dev, (uint8_t *)&ep_mgmt->setup_packet);
+		ep_mgmt->new_setup = true;
+	} else if (status_code == USB_DC_EP_DATA_OUT) {
+		__ASSERT_NO_MSG(USB_EP_DIR_IS_OUT(ep_cur->ep_addr));
+
+		/* Update EP read FIFO */
+		numaker_usbd_ep_fifo_update(ep_cur);
+
+		/* Need to clear NAK for next transaction */
+		ep_cur->nak_clr = false;
+	} else if (status_code == USB_DC_EP_DATA_IN) {
+		__ASSERT_NO_MSG(USB_EP_DIR_IS_IN(ep_cur->ep_addr));
+
+		/* Update EP write FIFO */
+		numaker_usbd_ep_fifo_update(ep_cur);
+
+		/* Need to clear NAK for next transaction */
+		ep_cur->nak_clr = false;
+	}
+}
+
+/* Interrupt bottom half processing
+ *
+ * This thread is used to not run Zephyr USB device stack and callbacks in interrupt
+ * context. This is because callbacks from this stack may use mutex or other kernel functions
+ * which are not supported in interrupt context.
+ */
+static void numaker_usbd_msg_hdlr_thread_main(void *arg1, void *arg2, void *arg3)
+{
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	__ASSERT_NO_MSG(arg1);
+	struct usb_dc_numaker_device *dev = (struct usb_dc_numaker_device *)arg1;
+	struct usb_dc_numaker_data *data = dev->data;
+
+	struct m_numaker_usbd_msg msg;
+
+	while (true) {
+		if (k_msgq_get(&data->msgq, &msg, K_FOREVER)) {
+			continue;
+		}
+
+		if (msg.msg_type == NUMAKER_USBD_MSG_TYPE_SW_RECONN) {
+			/* S/W reconnect for error recovery */
+			m_numaker_usbd_lock(dev);
+			m_numaker_usbd_sw_reconnect(dev);
+			m_numaker_usbd_unlock(dev);
+		} else if (msg.msg_type == NUMAKER_USBD_MSG_TYPE_CB_STATE) {
+			/* Interrupt bottom half processing for bus reset */
+			if (msg.cb_state.status_code == USB_DC_RESET) {
```

Comment:
the indentation starts to be a bit deep here and the code hard to follow, I'd suggest adding either one function to handle a generic message, something like

```c
while (true) {
        if (k_msgq_get(&data->msgq, &msg, K_FOREVER)) {
                continue;
        }
        numaker_usbd_msg_handle(dev, msg);
        ...
```

or even handlers for the individual message types

Reply:
still relevant

Reply:
updated following handlers for the individual message types